pkg/virt-handler/vm.go

```diff
@@ -186,11 +186,83 @@ type virtLauncherCriticalNetworkError struct {
 
 func (e *virtLauncherCriticalNetworkError) Error() string { return e.msg }
 
-func (d *VirtualMachineController) startDomainNotifyPipe(vmi *v1.VirtualMachineInstance) (net.Listener, error) {
+func handleDomainNotifyPipe(domainPipeStopChan chan struct{}, ln net.Listener, virtShareDir string, vmi *v1.VirtualMachineInstance) {
+
+	fdChan := make(chan net.Conn, 100)
+
+	// Listen for new connections,
+	// Close listener and exit when stop encountered
+	go func(vmi *v1.VirtualMachineInstance, ln net.Listener, domainPipeStopChan chan struct{}) {
+		for {
+			select {
+			case <-domainPipeStopChan:
+				log.Log.Object(vmi).Infof("closing notify pipe listener for vmi")
+				ln.Close()
+				return
+			default:
+				fd, err := ln.Accept()
+				if err != nil {
+					log.Log.Reason(err).Error("Domain pipe accept error encountered.")
+					// keep listening until stop invoked
+					time.Sleep(1)
+				} else {
+					fdChan <- fd
+				}
+			}
+		}
+	}(vmi, ln, domainPipeStopChan)
+
+	// Process new connections
+	// exit when stop encountered
+	go func(vmi *v1.VirtualMachineInstance, fdChan chan net.Conn, domainPipeStopChan chan struct{}) {
+		for {
+			select {
+			case <-domainPipeStopChan:
+				return
+			case fd := <-fdChan:
+				go func(vmi *v1.VirtualMachineInstance) {
+					defer fd.Close()
+
+					// pipe the VMI domain-notify.sock to the virt-handler domain-notify.sock
+					// so virt-handler receives notifications from the VMI
+					conn, err := net.Dial("unix", filepath.Join(virtShareDir, "domain-notify.sock"))
+					if err != nil {
+						log.Log.Reason(err).Error("error connecting to domain-notify.sock for proxy connection")
+						return
+					}
+					defer conn.Close()
+
+					log.Log.Object(vmi).Infof("Accepted new notify pipe connection for vmi")
+					copyErr := make(chan error)
```

Comment:
I fear a buffer is needed here. Otherwise the second error will block  go-routines:

```
make(chan error, 2)
```

Reply:
you're right. i'll fix this